cmd/qat_plugin/qat_plugin.go

```diff
@@ -0,0 +1,358 @@
+	}
+	id := bytes.TrimSpace(devId)
+	idStr := strings.TrimPrefix(string(id), "0x")
+	return idStr, nil
+}
+
+//Binds the device where id is the pci address to the specified device driver
+func BindDevice(dpdkDriver string, id string) error {
+
+	devicePCIAddr := concatStr("0000:", id)
+	unbindKernelDevicePath := concatStr(pciDeviceDir, devicePCIAddr, driverUnbindSuffix)
+	bindDevicePath := concatStr(pciDriverDir, dpdkDriver, newIdSuffix)
+	devicePCIAddrBytes := []byte(devicePCIAddr)
+	vfdevID, err := getDeviceID(devicePCIAddr)
+	if err != nil {
+		glog.Error(err)
+		fmt.Printf("Cannot obtain the Device ID for this device")
+		return fmt.Errorf("Cannot obtain the Device ID for this device: %v\n", err)
+	}
+	vfBindStr := concatStr(vendorPrefix, vfdevID)
+	vfBindStrBytes := []byte(vfBindStr)
+	err = ioutil.WriteFile(unbindKernelDevicePath, devicePCIAddrBytes, 0644)
+	if err != nil {
+		glog.Error(err)
+		fmt.Printf("Unbinding from the kernel driver failed\n")
+		return fmt.Errorf("Unbinding from the kernel driver failed: %v\n", err)
+
+	}
+	// Unbinding from the kernel driver DONE
+	err = ioutil.WriteFile(bindDevicePath, vfBindStrBytes, 0644)
+	if err != nil {
+		fmt.Printf("Binding to the dpdk driver failed\n")
+		return fmt.Errorf("Binding to the dpdk driver failed: %v\n", err)
+	}
+	//Binding to the the dpdk driver DONE\n
+	return nil
+}
+func isValidKerneDriver(kernelvfDriver string) error {
+	switch kernelvfDriver {
+	case "dh895xccvf", "c6xxvf", "c3xxxvf", "d15xxvf":
+		return nil
+	default:
+		return fmt.Errorf("Please enter the kernel-vf-drivers flag correctly")
+	}
+}
+func isValidDpdkDeviceDriver(dpdkDriver string) error {
+	switch dpdkDriver {
+	case "igb_uio", "vfio-pci":
+		return nil
+	default:
+		return fmt.Errorf("Please enter the dpdk-driver flag correctly")
+	}
+}
+
+// Discovers all QAT devices available on the local node by querying PCI bus using lspci.
+func (dm *deviceManager) discoverQATs() (bool, error) {
+
+	var found bool = false
+	fmt.Println("Discovered Devices below:")
+	kernelvfDrivers := strings.Split(*kernelVfDrivers, ",")
+
+	for _, kernelvfDriver := range kernelvfDrivers {
+
+		err := isValidKerneDriver(kernelvfDriver)
+		if err != nil {
+			return found, fmt.Errorf("Error in user input for kernel VF Driver: %v", err)
+		}
+
+		err = isValidDpdkDeviceDriver(*dpdkDriver)
+		if err != nil {
+			return found, fmt.Errorf("Error in user input for DPDK Device Driver: %v", err)
+		}
+
+		files, err := ioutil.ReadDir(concatStr(pciDriverDir, kernelvfDriver))
+		if err != nil {
+			fmt.Errorf("Can't read sysfs for kernel vf driver %v: %v", kernelvfDriver, err)
+			continue
+		}
+
+		for n, file := range files {
+			if strings.HasPrefix(file.Name(), "0000:") {
+				vfpciaddr := strings.TrimPrefix(file.Name(), "0000:")
+				max, err := strconv.Atoi(*maxNumdevice)
+				if err != nil {
+					return found, fmt.Errorf("Error in getting maximum number of devices: %v", err)
+				}
```

Comment:
User input checks should be moved off this cycle and the best place to do it is the `main()` function before we do any serious work for security reasons.